lib/srv/db/proxyserver.go

```diff
@@ -0,0 +1,321 @@
+limitations under the License.
+*/
+
+package db
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/native"
+	"github.com/gravitational/teleport/lib/auth/proto"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/multiplexer"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/postgres"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// ProxyServer runs inside Teleport proxy and is responsible to accepting
+// connections coming from the database clients (via a multiplexer) and
+// dispatching them to appropriate database services over reverse tunnel.
+type ProxyServer struct {
+	// ProxyServerConfig is the proxy server configuration.
+	ProxyServerConfig
+	// middleware extracts identity information from client certificates.
+	middleware *auth.Middleware
+	// closeCtx is closed when the process shuts down.
+	closeCtx context.Context
+	// log is used for logging.
+	log logrus.FieldLogger
+}
+
+// ProxyServerConfig is the proxy configuration.
+type ProxyServerConfig struct {
+	// AuthClient is the authenticated client to the auth server.
+	AuthClient *auth.Client
+	// AccessPoint is the caching client connected to the auth server.
+	AccessPoint auth.AccessPoint
+	// Authorizer is responsible for authorizing user identities.
+	Authorizer auth.Authorizer
+	// Tunnel is the reverse tunnel server.
+	Tunnel reversetunnel.Server
+	// TLSConfig is the proxy server TLS configuration.
+	TLSConfig *tls.Config
+}
+
+// CheckAndSetDefaults validates the config and sets default values.
+func (c *ProxyServerConfig) CheckAndSetDefaults() error {
+	if c.AccessPoint == nil {
+		return trace.BadParameter("missing AccessPoint")
+	}
+	if c.AuthClient == nil {
+		return trace.BadParameter("missing AuthClient")
+	}
+	if c.Authorizer == nil {
+		return trace.BadParameter("missing Authorizer")
+	}
+	if c.Tunnel == nil {
+		return trace.BadParameter("missing Tunnel")
+	}
+	if c.TLSConfig == nil {
+		return trace.BadParameter("missing TLSConfig")
+	}
+	return nil
+}
+
+// NewProxyServer creates a new instance of the database proxy server.
+func NewProxyServer(ctx context.Context, config ProxyServerConfig) (*ProxyServer, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	server := &ProxyServer{
+		ProxyServerConfig: config,
+		middleware: &auth.Middleware{
+			AccessPoint: config.AccessPoint,
+		},
+		closeCtx: ctx,
+		log:      logrus.WithField(trace.Component, "db:proxy"),
+	}
+	server.TLSConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	server.TLSConfig.GetConfigForClient = getConfigForClient(
+		server.TLSConfig, server.AccessPoint, server.log)
```

Comment:
nit: don't wrap here

Reply:
Not sure what you mean, don't wrap the line, or something else?